src/VectorizeLoops.cpp

```diff
@@ -212,6 +213,146 @@ class RewriteAccessToVectorAlloc : public IRMutator {
+            expr = Call::make(op->type, Call::predicated_load, {src, vector_predicate}, Call::Intrinsic);
+            vectorized = true;
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+    void visit(const Store *op) {
+        valid = valid && should_predicate_store_load(op->value.type().bits());
+        if (!valid) {
+            stmt = op;
+            return;
+        }
+        if (!op->index.type().is_scalar()) {
+            internal_assert(op->index.type().lanes() == lanes);
+            internal_assert(op->value.type().lanes() == lanes);
+            Expr value = mutate(op->value);
+            Expr dest = Call::make(Handle().with_lanes(lanes), Call::address_of,
+                                   {Load::make(op->value.type(), op->name, op->index, BufferPtr(), op->param)},
+                                   Call::Intrinsic);
+            stmt = Evaluate::make(Call::make(value.type(), Call::predicated_store,
+                                             {dest, vector_predicate, value},
+                                             Call::Intrinsic));
+            vectorized = true;
+        } else if (expr_uses_var(op->index, var)) {
+            Expr value = Broadcast::make(op->value, lanes);
+            Expr index = Broadcast::make(op->index, lanes);
+            Expr dest = Call::make(Handle().with_lanes(lanes), Call::address_of,
+                                   {Load::make(value.type(), op->name, index, BufferPtr(), op->param)},
+                                   Call::Intrinsic);
+            stmt = Evaluate::make(Call::make(value.type(), Call::predicated_store,
+                                             {dest, vector_predicate, mutate(value)},
+                                             Call::Intrinsic));
+            vectorized = true;
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+    Expr merge_predicate(Expr pred, Expr new_pred) {
+        if (pred.type().lanes() == new_pred.type().lanes()) {
+            Expr res = simplify(pred && new_pred);
+            return res;
+        }
+        valid = false;
+        return pred;
+    }
+
+    void visit(const Call *op) {
+        if (!valid) {
+            IRMutator::visit(op);
+            return;
+        }
+        if (op->is_intrinsic(Call::predicated_store) || op->is_intrinsic(Call::predicated_load)) {
+            vector<Expr> new_args(op->args);
+            new_args[1] = merge_predicate(op->args[1], vector_predicate);
+            if (!valid) {
+                expr = op;
+                return;
+            }
+            expr = Call::make(op->type, op->name, new_args, op->call_type,
+                              op->func, op->value_index, op->image, op->param);
+            vectorized = true;
+        } else {
+            valid = valid && op->is_pure();
+            expr = op;
+        }
+    }
+
+public:
+    // Should only attempt to predicate store/load if the lane size is
```

Comment:
This comment seems misplaced. Should it be further up on the should_... function?